cookbooks/aws-parallelcluster-slurm/files/default/head_node_checks/check_cluster_ready.py

```diff
@@ -0,0 +1,160 @@
+    #   }
+    # }
+    items_by_id = {item["Id"]["S"]: item["Data"]["M"] for item in items}
+
+    for instance_id in instance_ids:
+        key = CLUSTER_CONFIG_DDB_ID.format(instance_id=instance_id)
+        data = items_by_id.get(key)
+        if data is None:
+            logger.warning("Missing cluster config record for instance %s", instance_id)
+            missing.append(instance_id)
+            continue
+        cluster_config_version = data.get("cluster_config_version", {}).get("S")
+        if cluster_config_version is None:
+            logger.warning("Missing cluster config version in data for instance %s", instance_id)
+            incomplete.append(instance_id)
+            continue
+        if cluster_config_version != expected_config_version:
+            logger.warning(
+                "Wrong cluster config version for instance %s; expected %s, but got %s",
+                instance_id,
+                expected_config_version,
+                cluster_config_version,
+            )
+            wrong.append(instance_id)
+
+    return missing, incomplete, wrong
+
+
+def check_compute_nodes_config_version(cluster_name: str, table_name: str, expected_config_version: str, region: str):
+    """
+    Verify that every compute node in the cluster has deployed the expected config version.
+
+    The verification is made by checking the config version reported by compute nodes on the cluster DDB table.
+    A RuntimeError exception is raised if the check fails.
+    The function is retried and the wait time is expected to be in the interval (cfn_hup_time, 2*cfn_hup_time),
+    where cfn_hup_time is the wait time for the cfn-hup daemon (as of today it is 120 seconds).
+
+    :param cluster_name: name of the cluster.
+    :param table_name: DDB table to read the deployed config version from.
+    :param expected_config_version: expected config version.
+    :param region: AWS region name (eg: us-east-1).
+    :return: None
+    """
+    logger.info(
+        "Checking that cluster configuration deployed on compute nodes for cluster %s is %s",
+        cluster_name,
+        expected_config_version,
+    )
+
+    for instance_ids in list_cluster_instance_ids_iterator(
+        cluster_name=cluster_name,
+        node_type=["Compute"],
+        instance_state=["running"],
+        region=region,
+    ):
+        n_instance_ids = len(instance_ids)
+
+        if not n_instance_ids:
+            logger.warning("Found empty batch of compute nodes: nothing to check")
+            continue
+
+        logger.info("Found batch of %s compute node(s): %s", n_instance_ids, instance_ids)
+
+        items = get_cluster_config_records(table_name, instance_ids, region)
+        logger.info("Retrieved %s DDB item(s): %s", len(items), items)
+
+        missing, incomplete, wrong = _check_cluster_config_items(instance_ids, items, expected_config_version)
+
+        if missing or incomplete or wrong:
+            raise CheckFailedError(
```

Comment:
what about logging the error here rather than having a warning line in the log for each single failure (like we're doing in the _check_cluster_config_items)

Reply:
Goo point. I'll remove the single warnings, but it's not needed to add a new collective error message because the error message is included as the CheckFailedError message, which is logged upstream.